builder/dockerfile/shell_parser.go

```diff
@@ -166,15 +166,28 @@ func (sw *shellWord) processStopOn(stopChar rune) (string, []string, error) {
 func (sw *shellWord) processSingleQuote() (string, error) {
 	// All chars between single quotes are taken as-is
 	// Note, you can't escape '
+	//
+	// From the "sh" man page:
+	// Single Quotes
+	//   Enclosing characters in single quotes preserves the literal meaning of
+	//   all the characters (except single quotes, making it impossible to put
+	//   single-quotes in a single-quoted string).
+
 	var result string
 
 	sw.scanner.Next()
 
 	for {
 		ch := sw.scanner.Next()
-		if ch == '\'' || ch == scanner.EOF {
+
+		if ch == scanner.EOF {
+			return "", fmt.Errorf("Unexcepted end of statement while looking for matching single-quote")
```

Comment:
minor: the go style guide suggests errors should always start with a lowercase letter so that error messages look better when wrapped by other errors.